### ISSUE TITLE
"Too much recursion" when using "when" instead of if

### ISSUE THREAD

**Issue by lambdor** [faylang/fay#423]
The following code works (except that the first putStrLns are swallowed). However when I comment the "if" line and uncomment the "when" line I get a "too much recursion"-error and the program halts prematurely.

```
main :: Fay ()
main = do
    putStrLn "beg"
    loop
    putStrLn "end"
    where
        now :: Fay Double
        now = ffi "window.performance.now()"
        loop = do
            t <- now
            putStrLn . show . floor $ t
            if t < 500 then loop else return ()
--            when (t < 500) loop
```

**Comment by ryachza** [comment]
I don't believe this is resolvable via tail call elimination since `when` is declared as `Bool -> Fay a -> Fay ()` making it equivalent to `if t < 500 then loop >> return () else return ()` in this case.

Is updating `when` to `Bool -> Fay () -> Fay ()` matching `base` an option? I presume it was declared the way it is to avoid the need for explicit `void` in some cases?

**Comment by bergmark** [comment]
Sorry for the late reply.

Good catch! I think this is simply a mistake, I don't want fay-base to differ from base unless necessary.

**Comment by bergmark** [comment]
This was fixed by #428
